python_scripts/parameter_tuning_nested.py

```diff
@@ -69,56 +66,155 @@
+from sklearn.model_selection import train_test_split
+
+data_train, data_test, target_train, target_test = train_test_split(
+    data, target, test_size=0.2, random_state=42
+)
+
+model_grid_search.fit(data_train, target_train)
+accuracy = model_grid_search.score(data_test, target_test)
+print(f"Accuracy on test set: {accuracy:.3f}")
 
 # %% [markdown]
-# The hyperparameters on each fold are potentially different since we nested
-# the grid-search in the cross-validation. Thus, checking the variation of the
-# hyperparameters across folds should also be analyzed.
+# In the code above, the selection of the best hyperparameters was done only on
+# the train set. Then, we evaluated the generalization performance of our
+# tuned model on the left out test set.
+#
+# However, this evaluation only provides us a single point estimate of the
+# generalization performance. As recall at the beginning of this notebook,
+# it is beneficial to have an uncertainty measure of the generalization
+# performance. Therefore, we should instead use a cross-validation for this
+# evaluation.
+#
+# This pattern is called *nested cross-validation*. We use an inner
+# cross-validation for the selection of the hyperparameters and an outer
+# cross-validation for the evaluation of generalization performance of the
+# refitted tuned model.
+#
+# In practice, we only need to embed the grid-search in the function
+# `cross-validate` to perform such evaluation.
+
+# %%
+cv_results = cross_validate(
+    model_grid_search, data, target, cv=5, n_jobs=2, return_estimator=True
+)
 
 # %%
-for fold_idx, estimator in enumerate(cv_results["estimator"]):
-    print(f"Best parameter found on fold #{fold_idx + 1}")
-    print(f"{estimator.best_params_}")
+cv_results = pd.DataFrame(cv_results)
+print(
+    "Generalization score with hyperparameters tuning:\n"
+    f"{cv_results['test_score'].mean():.3f} +/- {cv_results['test_score'].std():.3f}"
+)
 
 # %% [markdown]
-# Obtaining models with unstable hyperparameters would be an issue in practice.
-# Indeed, it would become difficult to set them.
+# In this case, we obtain a distribution of scores and therefore, we can
+# apprehend the variability of the generalization performance.
+#
+# In addition, passing the parameter `return_estimator=True`, we can check the
+# value of the best hyperparameters for each fold of the outer
+# cross-validation.
+
+# %%
+for cv_fold, estimator_in_fold in enumerate(cv_results["estimator"]):
+    print(
+        f"Best hyperparameters for fold #{cv_fold + 1}:\n"
+        f"{estimator_in_fold.best_params_}"
+    )
 
 # %% [markdown]
+# Such an analysis allows to see if only a single set of hyperparameters is
+# sufficient to obtain a good generalization performance or if there is some
+# variation in the hyperparameters used.
+#
+# Obtaining models with unstable hyperparameters would be an issue in practice.
+# Indeed, it would become difficult to set them.
```

Comment:
Maybe the following is more operation:

```suggestion
# It is interesting to see whether the hyper-parameter tuning procedure always select
# similar values for the hyper-parameters. If its the case, then all is fine. It means that
# we can deploy a model fit with those hyper-parameters and expect that it will have
# an actual predictive performance close to what we measured in the outer
# cross-validation.
#
# But it is also possible that some hyper-parameters do not matter at all, and as a
# result in different tuning sessions give different results. In this case,
# any value will do. This can typically be confirmed by doing a parallel coordinate plot
# of the results of a large hyper-parameter search as seen in the exercises.
#
# From a deployment, one could also chose to deploy all the models found by the outer
# cross-validation loop and make them vote to get the final predictions. However this
# can cause operational problems because it uses more memory and makes
# computing prediction slower resulting in a higher computational resource usage
# per prediction. 
```